### ISSUE TITLE
7.26.0 Upgrade Issues - InjectionStatus PreviouslyPartiallyInjected

### ISSUE THREAD

**Issue by rimbesat** [DataDog/chaos-controller#748]
**Describe the bug**
Hi team, We recently tried to upgrade from version 7.13.1 to 7.26 and we are facing issues trying to achieve the same behaviour. We previously reached out to you about [issues with dynamic targeting](https://github.com/DataDog/chaos-controller/issues/744), we upgraded to your newest release and we are hoping to get some further guidance.
We wanted to understand the expected behaviour of the Injection status. For example, when running cpu pressure, the final status ends up being PreviouslyPartiallyInjected despite the experiment injecting all the desired pods. 

**To Reproduce**

1. This is an example of a cpu-pressure experiment
```
apiVersion: chaos.datadoghq.com/v1beta1
kind: Disruption
metadata:
  name: cpu-pressure-pod-rohab
  namespace: chaos-engineering-framework
  annotations:
    chaos.datadoghq.com/environment: dev
spec:
  level: pod
  duration: 6m
  selector:
    app.kubernetes.io/name: "fault-injection-showcase"
    app: "fault-injection-showcase--plugin-master"
  count: 100%
  staticTargeting: true
  cpuPressure: 
    count: 100%
```
The cpu pressure correctly identify targets and injects the failure. We believe the logs attached below for cpu-pressure show that the targeted pods are not seen as ready and later go into a terminated state 
Flow:
1.Disruption Created
2. 2 Tagets identified
3. 2 chaos pods created.
4. No Status straight to "PartiallyInjected"
5. 0 pods ready, desired 2
6. 1 chaos pod ready, desired is 2. Status "PartiallyInjected"
7. 2 chaos pods ready, desired is 2. Status "Injected"
8. 1 Chaos pod goes in to a not Ready status
9. 1 chaos pod ready, desired is 2. Status "PartiallyInjected" from "Injected"
10, Error:
```
"caller": "controllers/disruption_controller.go:148",
  "message": "a retryable error occurred in reconcile loop",
  "disruptionName": "cpu-pressure-pod-rohab",
  "disruptionNamespace": "chaos-engineering-framework",
  "error": "error handling chaos pods termination: Operation cannot be fulfilled on disruptions.chaos.datadoghq.com \"cpu-pressure-pod-rohab\": the object has been modified; please apply your changes to the latest version and try again"
```
11. 0 chaos pod ready, desired is 2, Termination 1 Status "PartiallyInjected" from "PausedPartiallyInjected"
12. 0 chaos pod ready, desired is 2, Termination 2 Status "PausedPartiallyInjected" from "PreviouslyPartiallyInjected"

**Expected behavior**
We expect the experiment to finish on PreviouslyInjected if all the desired pods are injected.

**Logs**
[chaos-controller-27-06.txt](https://github.com/DataDog/chaos-controller/files/12185506/chaos-controller-27-06.txt)
[cpu_pressure_partially_injection_debug_chaos_pod_log.txt](https://github.com/DataDog/chaos-controller/files/12185507/cpu_pressure_partially_injection_debug_chaos_pod_log.txt)

Describe on pods on cpu pressure(from a different experiment time but same experiment manifest)
[injector_pod_describe.txt](https://github.com/DataDog/chaos-controller/files/12185558/injector_pod_describe.txt)
[target_pod_describe.txt](https://github.com/DataDog/chaos-controller/files/12185559/target_pod_describe.txt)

If you have any ideas. please let us know. 

**Environment:**
Kubernetes version: v1.24.15-eks-a5565ad
Controller version: 7.26.0
Cloud provider (or local): AWS EKS
Base OS for Kubernetes: Amazon Linux: 5.4.242-156.349.amzn2.x86_64

**Comment by ptnapoleon** [comment]
I'll take a look tomorrow, thanks for the issue report

**Comment by ptnapoleon** [comment]
Indeed, every state transition you listed was valid (at least as written in the code) and expected. The terminating status PreviouslyPartiallyInjected only guarantees that at the time of termination, we had a status of ~PartiallyInjected. We don't attempt to track degradations from Injected to PartiallyInjected or NotInjected.

I'm not sure why we went from 2 chaos pods ready to just 1, from the attached logs.

Is the root of the issue that you check the final status of a Disruption to determine its success, and ending up with PreviouslyPartiallyInjected instead of PreviouslyInjected leaves it unclear if the disruption worked or not?

I don't think we want to make the list of possible states longer, and while we could track if the Disruption had _ever_ reached Injected, we can't guarantee that even if we did reach a time where all chaos pods are ready, that we would reconcile at the correct time to reflect that in the status.

**Comment by expFlower** [comment]
Thanks for the responses both:

Firstly @ptnapoleon, the first point is attempting to understand the transition from 2 chaos pods read to just 1. This was something we were struggling to explain through the logs from either the manager or injector along with inspecting the cluster itself. 

You are indeed correct:
> Is the root of the issue that you check the final status of a Disruption to determine its success, and ending up with PreviouslyPartiallyInjected instead of PreviouslyInjected leaves it unclear if the disruption worked or not?

We have built integration around the chaos controller, via a control plane and CICD platform to reduce the friction for users to onboard, execute and monitor experiments. The end status of PreviouslyPartiallyInjected in this particular situation raises questions that can't be answered at this moment in time.

I'll setup the controller locally and point it to the cluster to where this is happening to try and grab more information to debug.

**Comment by ptnapoleon** [comment]
I've double checked the code for cpu_pressure, and there doesn't appear to be any situation in which an injector pod would mark itself as Not Ready after having been Ready previously.

Looks like `chaos-cpu-pressure-pod-rohab-q4q6j` is the chaos pod that transitions out of Ready, do you have the logs from that pod? (or a pod with matching behavior) It isn't clear from your attached log which injector pod it belongs to. It could also be helpful if you attached a describe of an injector pod while it was in a NotReady state (after having been Ready)

**Comment by expFlower** [comment]
Hi Philip, thanks for the response,

I've managed to grab both a describe pod and the pod logs from a chaos pod that marks itself not ready and the logs from the manager during the period.

We can see the readiness probe from the chaos injector pod failing after 7m30s and are trying to trace the reason why this is occurring. It's very difficult to grab the pod describe at the exact time.

Please note that this behaviour is transient, in that if we run the same exact experiment in the same environment we don't always see this behaviour.
 
[chaos_manager_log.txt](https://github.com/DataDog/chaos-controller/files/12261080/chaos_manager_log.txt)
[chaos_pod_container_exit_describe.txt](https://github.com/DataDog/chaos-controller/files/12261081/chaos_pod_container_exit_describe.txt)
[chaos_pod_container_exit_log.txt](https://github.com/DataDog/chaos-controller/files/12261082/chaos_pod_container_exit_log.txt)
[chaos_disruption_describe.txt](https://github.com/DataDog/chaos-controller/files/12261196/chaos_disruption_describe.txt)

**Comment by ptnapoleon** [comment]
It's a remarkably small patch, but I think https://github.com/DataDog/chaos-controller/compare/phil/issue-748?expand=1 might help? 

Assuming the clocks are synced, I'm seeing that within the same second (about 700ms apart), the chaos pod completes its lifecycle and stops. The disruption notices that its remaining duration has hit 0 seconds and does a `getChaosPods` followed by `disruptionTerminationStatus`. It's in this second call that we are downgrading the state from Injected to PartiallyInjected. But if the chaos pod is right to have evaluated its duration as complete (a fact I'm taking for granted), then I think we're hitting a one second long race condition where disruption_controller.go's Reconcile is considering the disruption expired, because `remainingDuration <= 0`, but `disruptionTerminationStatus` is not, because `remainingDuration < 0` is the check done there, and `remainingDuration` is measured in Seconds, for both. Let me know if you need me to push an image of this branch to dockerhub

**Comment by expFlower** [comment]
Thanks for the update. i've got the controller running locally in my IDE pointing to the cluster for debugging. I'll use the branch you've posted above and run some validations against this and report the findings to the thread.

**Comment by expFlower** [comment]
Initial verification of this patch has been positive, would be possible to work this in to an image from a branch or work on a new release with this patch. If you need anything further from us please shout. Thanks

**Comment by ptnapoleon** [comment]
Absolutely, I'll open a PR today and try to get it out asap

**Comment by ptnapoleon** [comment]
I'm building a 7.26.1 tag right now. It has exactly one commit on top of 7.26.0, that should be available within the hour

**Comment by expFlower** [comment]
Morning @ptnapoleon, looks like the build had an issue any chance you could take a look. Thanks

**Comment by expFlower** [comment]
Hi @ptnapoleon, sadly when we starting running the new 7.26.1 version in our dev cluster the problem is still there. That is that the injector pods go in to a not ready state and cause the disruption to end in a previouslyPartiallyInjected state.

Please advise what diags / logs etc we can provide in additional to what's already in this ticket. 

Thanks
David

**Comment by ptnapoleon** [comment]
Could you provide the same logs as before, just with 7.26.1? So one from the chaos-controller's manager container during the lifecycle of the disruption in question, and one from an injector pod that went NotReady from that same disruption

**Comment by expFlower** [comment]
Sure thing:

[chaos_describe_distruption.txt](https://github.com/DataDog/chaos-controller/files/12322281/chaos_describe_distruption.txt)
[chaos_describe_pod.txt](https://github.com/DataDog/chaos-controller/files/12322282/chaos_describe_pod.txt)
[chaos-controller-bff957b6-4vscz.txt](https://github.com/DataDog/chaos-controller/files/12322283/chaos-controller-bff957b6-4vscz.txt)
[chaos-cpu-pressure-pod-debug-d58tp.txt](https://github.com/DataDog/chaos-controller/files/12322286/chaos-cpu-pressure-pod-debug-d58tp.txt)

Thanks

**Comment by ptnapoleon** [comment]
There's a ~600ms disagreement between the chaos pod and the chaos-controller as to when the duration was meant to expire. We'll discuss internally as how to best handle that.

**Comment by ptnapoleon** [comment]
We've agreed on a fix, but due to PTO on the team, we likely won't get to it for a few weeks. If you have an urgent need, reach out to @Devatoria. Thanks for your patience

**Comment by expFlower** [comment]
Thanks for the update @ptnapoleon, it's not urgent as we can explain the condition to our customers should this condition occur. Looking forward to the fix.

**Comment by ptnapoleon** [comment]
Sorry for the delay, I was out for a month. We discussed, and we aren't currently trying to guarantee the behavior you're hoping for. Doing so would require more complicated termination logic than we want to handle. We're hoping to compromise with a "usually right" fix, that in the absence of clock drift >1 second between the controller and injector pods, should give the behavior you expect. At the very least, it should be no worse than it currently is.

I'm rebasing it on top of https://github.com/DataDog/chaos-controller/pull/764 and will test the patch on our side to confirm it performs as expected.

**Comment by ptnapoleon** [comment]
Merging the PR auto-closed this, but feel free to re-open if the 8.1.2 release (being built right now) still has the same problem for you
